### ISSUE TITLE
Loss explosion during training

### ISSUE THREAD

**Issue by Standchen** [SWivid/F5-TTS#832]
### Checks

- [x] This template is only for bug reports, usage problems go with 'Help Wanted'.
- [x] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [x] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [x] I confirm that I am using English to submit this report in order to facilitate communication.

### Environment Details

- OS: Ubuntu 22.04 LTS
- GPU: 8 x NVIDIA A100
- Python 3.10
- CUDA 12.1
- torch==2.3.1+cu121 / accelerate==1.2.0

### Steps to Reproduce

Follow the instruction and run the training.

### ✔️ Expected Behavior

The loss should gradually decrease and plateau around 0.6.
_(While this plateau doesn’t indicate full convergence, it aligns with the author's explanations in other issues, suggesting continued model improvement.)_

### ❌ Actual Behavior

The loss **explodes** unpredictably at random points.

- I've tried multiple training runs, but they've all failed due to loss explosion.
- Resuming from a stable checkpoint (before any issues) with different seeds allows training to continue for some steps, but the loss eventually explodes again.
- The dataset has been verified to be clean and free of any issues.
- In some cases, the loss goes `NaN` after the explosion.

I’ve attached loss plots below showing stable behavior around 0.6 before sudden divergence.

I have no idea why this is happening, and would greatly appreciate any insights or suggestions.

<img width="1249" alt="Image" src="https://github.com/user-attachments/assets/95bb05e7-cb54-463d-8f69-b485c95d40ef" />
<img width="1248" alt="Image" src="https://github.com/user-attachments/assets/250cff50-aa88-4dca-b69b-312fc98154cf" />
<img width="473" alt="Image" src="https://github.com/user-attachments/assets/eb71bd42-f2c9-41c2-96dc-44eef42366ca" />

**Comment by SWivid** [comment]
how is your training config, e.g. mix precision choice, the duration interval of audios

**Comment by Standchen** [comment]
@SWivid 

Hello. Thank you for your reply!

This is `accelerate` settings I used. I enabled **_mixed precision_** with `bf16`.

```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: MULTI_GPU
downcast_bf16: 'no'
enable_cpu_affinity: false
gpu_ids: all
machine_rank: 0
main_training_function: main
mixed_precision: 'bf16'
num_machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

Below is the length distribution of my dataset.

![Image](https://github.com/user-attachments/assets/f1e3accf-57f7-465a-a414-0fa0ab1d63f3)

**Comment by SWivid** [comment]
Got your setting.

Based on some valuable feedbacks from another person, very short audio samples could give poisonous loss values under bf16 precision that gradually accumulate to break the training.

From our own experiments, except the early triage with bigvgan as vocoder, we always train with fp16 and bug-free.
And for reference, we mostly use Emilia (3s~30s) and LibriTTS for ablation.

**Comment by SWivid** [comment]
> should I use fp16 instead of bf16?

I would recommand fp16.

If you are not in a hurry, could wait for few days. We will add some fixes to current version, e.g. #793

**Comment by Standchen** [comment]
Based on your valuable feedback,
I’ve tried using `fp16` instead of `bf16`, but the training remains unstable (with the loss eventually becoming `NaN` this time.)

Now, I’d like to filter out short samples from my dataset. Could you advise on the audio length _threshold_ to avoid _"poisonous loss values"_?

**Comment by SWivid** [comment]
> And for reference, we mostly use Emilia (3s~30s) and LibriTTS for ablation.

Maybe could start from a workable dataset, and switch to your own corpus, if don't want to much samples filtered out?
Or does it make sense to concate short samples with same speaker

**Comment by ZhikangNiu** [comment]
Is your dataset public? If so, we can reproduce your exp to check this question

**Comment by Standchen** [comment]
@SWivid @ZhikangNiu 
I’ve already concatenated short audios from the same speakers to enrich my dataset (though those concatenated samples does not appear in the histogram I've attached), but I appreciate the suggestion!

Unfortunately, the dataset I’m using is not publicly available. However, I’ve run a training experiment using the **_Emilia_** dataset, as you've used for your experiments.
Additionally, I filtered out short samples _(< 3s)_ from my dataset and re-ran the training.

Below are the loss plots from both runs (**_Emilia_** vs. **_my filtered dataset_**) for comparison.
Although they only cover the early stage of training (_~23k steps_), I’d like to ask if these loss trends appear normal (in terms of stability, range, etc.)

Thanks again for your help!

![Image](https://github.com/user-attachments/assets/20158613-817b-48f1-91a1-7ac93d4f8dbd)
![Image](https://github.com/user-attachments/assets/13e041eb-8948-480c-a223-d636eb2a0de4)
![Image](https://github.com/user-attachments/assets/3a5e7728-eae9-4050-9c74-334d9bcf2e32)
![Image](https://github.com/user-attachments/assets/f8dbcfa5-472e-4e09-b8d8-77fdc1c24379)

**Comment by SWivid** [comment]
@Standchen Hi, it's normal, the mel loss bumps between 0.6 1.0 in our case too.

We are also looking into the loss issues, would recommand you to add `self.drop_last = True` to https://github.com/SWivid/F5-TTS/blob/09b478b7d73b54f1c5e5e90f5372126e6d128a74/src/f5_tts/model/dataset.py#L216 and see if help avoid NaN.

**Comment by awei669** [comment]
> > should I use fp16 instead of bf16?
> 
> I would recommand fp16.
> 
> If you are not in a hurry, could wait for few days. We will add some fixes to current version, e.g. [#793](https://github.com/SWivid/F5-TTS/issues/793)

Looking forward to the new version.

**Comment by SWivid** [comment]
> Looking forward to the new version.

@awei669 pushed, feel free to try~

**Comment by SynthAether** [comment]
@SWivid Thanks for the update. It's much appreciated. I was in the middle of training a model for a few days. Is the new update backward compatible with the saved checkpoint from the previous version of the code?

**Comment by SWivid** [comment]
@SynthAether we have ensured the backward compatibility with previous model, i.e. you could use src/f5_tts/configs/F5TTS_Base.yaml
But we would recommand to use F5TTS_v1_Base.yaml, which performs better~

**Comment by MuruganR96** [comment]
@SWivid, with backward compatibility with the previous model, can I continue the new changes (i.e., zero initialization for adaptive layer norm in DiT)?

**Comment by SWivid** [comment]
> Suppose how the model will perform? pleae tell me

https://swivid.github.io/F5-TTS_updates/ we got some random test, you could just test whatever you want with v1 model (model weight already pushed on huggingface/modelscope).

> with backward compatibility with the previous model, can I continue the new changes (i.e., zero initialization for adaptive layer norm in DiT)?

zero initialization is literally done at initialization when just start the training, but other changes force a rerun if you want to make the update.

**Comment by SWivid** [comment]
will close this issue, feel free to open if any updates

**Comment by rasenganai** [comment]
@SWivid, I'm facing a similar problem, im trying to finetune the base_v1_model on custom dataset comprises of IndicF5 model dataset + custom 1000 hours of data.
Overall it is a multilingual model with 6k hours over 11 languages, with each language containing good duration distribution 3-30s.
Im fine-tuning on 8h100 using fp16 and same default hyperparameters, and the model loss goes to nan after 550k iterations.
Any suggestions to look for this bug?
Thanks

**Comment by SWivid** [comment]
@rasenganai Based on our training experience, filtering out dirty data pairs is of top priority to avoid nan issue.

**Comment by rasenganai** [comment]
@SWivid Should i track loss for each data point and eliminate the outliers? 
What do you think of this approach?

**Comment by SWivid** [comment]
> Should i track loss for each data point and eliminate the outliers?

@rasenganai it may not work. Cuz with in-filling training, sometimes the flawed part of a data pair may be mask, sometimes not.
So if you launch two training with exact same config, one may face nan issue, one may not.

Personally, I think it is necessary to directly filter out bad samples before training process.
The specific method will depend on the nature of your data, e.g. you can use ASR, etc.

**Comment by rasenganai** [comment]
@SWivid Thanks for this insight, I'll try force alignment now to remove the dirty pairs, i have already made sure in my data cleaning pipeline to do various verification (Like multiple ASR models ensembling, speaker diarization etc).
The model upto 500k iterations is working good enough!
I'd run it on Small model upto 1M iterations first to see if it explodes.
